### PR TITLE
Prepare to 0.2.0 release

### LINKED ISSUE #169: Remove privnet genesis templates
# Rationale

We have a set of static validator wallets for privnet, they likely will never be regenerated. At the same time it's just useless extra work to keep the genesis templates up-to-date and relevant wrt privnet genesis configuration files.

# Implementation

Refactor Makefile, remove privnet wallets generation part from it. Remove privnet genesis template files.

### LINKED ISSUE #175: Add documentation about NeoX node start
# Rationale

Although almost no changes were made in the node configuration flags, we're currently missing NeoX node start guide.

# Implementation

We need to adjust the README.md to reflect the relevant NeoX node information. We also need to add a documentation about how to start the NeoX node and join the testnet (and mainnet in future).

### LINKED ISSUE #224: Ensure Solidity compiler ^0.8.25 is compatible with dBFT-supported hardforks
Ref. https://github.com/bane-labs/go-ethereum/pull/208#discussion_r1620928382, ref. https://github.com/bane-labs/go-ethereum/issues/206. We can't update compiler version without the dBFT compatibility check.

### LINKED ISSUE #226: [Documentation] Integrate Treasury Contract in Neo X Architecture
The treasury introduced in #184 should be added to the architecture in the `/contracts/README.md`.

### LINKED ISSUE #231: Allocate empty upgradable system contracts in the genesis file for future usage
# Rationale

Currently, we already have several system contracts like GovReward, Governance and Policy, and they all have pre-allocated addresses for the proxy storage. But when we want to add new contracts like Treasury, it's hard to give them similar addresses since we can't change genesis allocation for a deployed network. So they may have to be deployed to normal smart contract addresses.

# Implementation

Since our system contract are upgradable and controlled by governance, we can allocate enough empty contracts in the genesis file, so that we can use these addresses in the future.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Depends on #230, #166.

Close #169, close #226, close #224, close #231, close #175.

This PR needs to be merged before T4-compatible release.

### CODE REVIEW

File: config/genesis_testnet.json
Comment (AnnaShaleva) [inline config/genesis_testnet.json]: @mialbu, @gsmachado, @steven1227 for T4 all native Bridge contract funds are moved to native Treasury contract. Do we need to supply Bridge contract with at least *some* initial amount of GAS in the genesis allocations?

File: config/genesis_testnet.json
Comment (AnnaShaleva) [inline config/genesis_testnet.json]: @mialbu, @gsmachado, @steven1227 what is the desired balance of native Treasury contract for T4?

File: config/genesis_testnet.json
Comment (AnnaShaleva) [inline config/genesis_testnet.json]: @chenquanyu, @txhsl, @mialbu I'm not sure how to get the initial storage items for Treasury proxy contract. Could you please, explain or provide them?

File: config/genesis_testnet.json
Comment (AnnaShaleva) [inline config/genesis_testnet.json]: @chenquanyu, @txhsl all our native contracts are upgradable and supposed to have the same Proxy code, right? I noticed that some of our Proxy contracts have different code than the others. It's probably due to different compiler versions or something like that, but we need to unify this point. What is the desired code for the Proxy contracts?

File: config/genesis_testnet.json
Comment (AnnaShaleva) [inline config/genesis_testnet.json]: @chenquanyu, @txhsl, @mialbu could you please check that this Treasury implementation code matches the expectations? I compiled it using our hardhat config from `./contracts/hardhat.config.ts` with `evmVersion: "shanghai"`.

File: config/genesis_testnet.json
Comment (AnnaShaleva) [inline config/genesis_testnet.json]: @chenquanyu, @txhsl I need your help in storage allocations for systemcontracts stubs. Also, please check that Proxy/implementation code matches the expectations.

File: contracts/solidity/interfaces/IStub0.sol
Comment (AnnaShaleva) [inline contracts/solidity/interfaces/IStub0.sol]: @txhsl, @chenquanyu is this contract a valid Eth contract? We need to ensure that there will be no problem once we need to upgrade it.

**Review by AnnaShaleva — COMMENTED** [review]
Please, consider answering these questions, it will help to finish this PR.

File: contracts/solidity/interfaces/IStub0.sol
Comment (txhsl) [reply contracts/solidity/interfaces/IStub0.sol]: It is, but to make the upgrade possible, we need it to inherit `GovProxyUpgradeable.sol`, and contain the following codes. :)

https://github.com/bane-labs/go-ethereum/blob/7017f7aa5fe678b90496564ebb0d155d51095976/contracts/solidity/Governance.sol#L15

And.

https://github.com/bane-labs/go-ethereum/blob/7017f7aa5fe678b90496564ebb0d155d51095976/contracts/solidity/Governance.sol#L69-L87

The `SELF` address is the corresponding implementation address, which is **different** in each implementation file.

File: config/genesis_testnet.json
Comment (txhsl) [reply config/genesis_testnet.json]: Only a key `0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc` with its value `0x1212100000000000000000000000000000000006`.

File: config/genesis_testnet.json
Comment (txhsl) [reply config/genesis_testnet.json]: Codes for addresses from `0x1212000000000000000000000000000000000001` to `0x1212000000000000000000000000000000000003` are provided by @chenquanyu and me, while from `0x1212000000000000000000000000000000000004` to `0x1212000000000000000000000000000000000005` are provided by AxLabs, so they could be different.

From my point of view, I will recommend the code of `0x1212000000000000000000000000000000000001`.

File: config/genesis_testnet.json
Comment (txhsl) [reply config/genesis_testnet.json]: The proxy code should be the same as other proxies', and the storage allocation needs only a key `0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc` with its corresponding implementation address, just like the Treasury allocates.

File: contracts/solidity/interfaces/IStub0.sol
Comment (chenquanyu) [reply contracts/solidity/interfaces/IStub0.sol]: Maybe we can write a new contract to store `SELF` in a specific storage key. Then we only need one contract for all the empty upgradable system contracts.

Comment (txhsl) [reply contracts/solidity/interfaces/IStub0.sol]: These codes and value is temporary usage, because UUPS proxy can't get correctly initialized if allocated in genesis file. After the first upgrade, new implementation contracts will get `natural` addresses and don't need these any more.

File: config/genesis_testnet.json
Comment (AnnaShaleva) [reply config/genesis_testnet.json]: Implementation storage item is added. @mialbu, could you please confirm that no additional storage items are needed?

File: config/genesis_testnet.json
Comment (AnnaShaleva) [reply config/genesis_testnet.json]: > I will recommend the code of 0x1212000000000000000000000000000000000001

Fixed, let's use it for new contracts.

File: config/genesis_testnet.json
Comment (AnnaShaleva) [reply config/genesis_testnet.json]: Fixed, could you check please?

File: contracts/solidity/interfaces/IStub0.sol
Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub0.sol]: > to make the upgrade possible, we need it to inherit GovProxyUpgradeable.sol

Oh gosh, I was just copy-pasting the Treasury contract and completely forgot that Treasury is not upgradeable. Fixed, @txhsl, could you please check one more time?

File: config/genesis_testnet.json
Comment (txhsl) [reply config/genesis_testnet.json]: LGTM.

File: contracts/solidity/interfaces/IStub0.sol
Comment (txhsl) [reply contracts/solidity/interfaces/IStub0.sol]: Yeh, if Treasury is not upgradeable, then things are similar with GovProxyAdmin at `0x1212000000000000000000000000000000000000`. There will not be an implementation address, but only the implementation code in `0x1212000000000000000000000000000000000006`, and no need to allocate `0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc`.

Comment (txhsl) [reply contracts/solidity/interfaces/IStub0.sol]: BTW, we have updated nearly most of the contracts (including the bridge), due to the audit and some new features, so their implementation codes need to be compiled and updated again as well.

Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub0.sol]: > Yeh, if Treasury is not upgradeable, then things are similar with GovProxyAdmin

Fixed, check it please.

> so their implementation codes need to be compiled and updated again as well.

Implementations are updated for GovernanceProxyAdmin, Governance, Policy and GovernanceReward contracts. But storage items are still not updated, I need your help to update storage items.

To update Bridge and BridgeManagement implementaitons I need the final version of these contracts for T4 from @mialbu.

File: config/genesis_testnet.json
Comment (AnnaShaleva) [reply config/genesis_testnet.json]: UPD: Treasury is not upgradeable, thus we use a single genesis allocation for it with `0x1212000000000000000000000000000000000006` address and an empty storage items set.

File: config/genesis_testnet.json
Comment (mialbu) [reply config/genesis_testnet.json]: Imo there should be some initial allocation to the bridge. Otherwise, we'd require a governance vote before the bridge can be used.

File: config/genesis_testnet.json
Comment (mialbu) [reply config/genesis_testnet.json]: I'm fixing some things from the audit still. It'd be great if we can use that new contract for the start of T4. The medium issues are already fixed (unmerged PRs), and there's one or two low severity issues that I'll implement tomorrow. After that I'll need to have another look at the storage entries anyway.

File: contracts/solidity/interfaces/IStub0.sol
Comment (mialbu) [reply contracts/solidity/interfaces/IStub0.sol]: > To update Bridge and BridgeManagement implementaitons I need the final version of these contracts for T4 from @mialbu.

https://github.com/bane-labs/go-ethereum/pull/236#discussion_r1662813824

Comment (txhsl) [reply contracts/solidity/interfaces/IStub0.sol]: >  But storage items are still not updated, I need your help to update storage items.

Since I made the contract storage compatible, we only need to add a new field for `candidateLimit`. Also, @steven1227 wants to increase both the `baseFee` and `minGasTipCap` to `20 Gwei`.
```json
"0x1212000000000000000000000000000000000002": {
      "storage": {
        "0x0000000000000000000000000000000000000000000000000000000000000002": "0x04a817c800",
        "0x0000000000000000000000000000000000000000000000000000000000000003": "0x04a817c800",
        "0x0000000000000000000000000000000000000000000000000000000000000004": "0x07d0",
        "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc": "0x1212100000000000000000000000000000000002"
      }
    },
```

File: contracts/solidity/interfaces/IStub1.sol
Comment (chenquanyu) [inline contracts/solidity/interfaces/IStub1.sol]: Why do we need interface and method for stub contracts?

File: config/genesis_testnet.json
Comment (AnnaShaleva) [reply config/genesis_testnet.json]: Yes, it's the same chicken-and-egg problem. So how much funds do you think we should give to the bridge?

File: config/genesis_testnet.json
Comment (AnnaShaleva) [reply config/genesis_testnet.json]: It's OK, just write back once the contracts are in their final state.

File: contracts/solidity/interfaces/IStub1.sol
Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub1.sol]: It's just a template for further extension, but if you'd like then I can remove interface and remove this method from this contract. Agree?

File: contracts/solidity/interfaces/IStub0.sol
Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub0.sol]: Updated. And just to clarify: we don't need to change storage items for Governance or GovernanceVote contracts, right? There were changes in these contracts, but these changes did not affect the initializable variables.

Comment (txhsl) [reply contracts/solidity/interfaces/IStub0.sol]: > There were changes in these contracts, but these changes did not affect the initializable variables.

Yes.

File: contracts/solidity/interfaces/IStub1.sol
Comment (chenquanyu) [reply contracts/solidity/interfaces/IStub1.sol]: Maybe we don't need so many new empty implementation contracts, we can just point all these proxy contracts to the same implementation contract.

Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub1.sol]: > I can remove interface and remove this method from this contract

Done.

> we can just point all these proxy contracts to the same implementation contract.

Can we? Implementation contract should be able to control the upgrade and all implementation contracts have different `SELF`:
```
address public constant SELF = 0x1212100000000000000000000000000000000009;
```

Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub1.sol]: And I don't think that it's correct to point to the shared implementation from different proxies because we may have problems once we're going to update the implementation for one of the proxies. It's not quite clear to me what will happen with the rest of proxies in this case.

Could you please clarify? May be I'm wrong.

Comment (chenquanyu) [reply contracts/solidity/interfaces/IStub1.sol]: > And I don't think that it's correct to point to the shared implementation from different proxies because we may have problems once we're going to update the implementation for one of the proxies. It's not quite clear to me what will happen with the rest of proxies in this case.
> 
> Could you please clarify? May be I'm wrong.

Nothing will happen to other proxies  when you upgrade one proxy, you just change the implementation slot value when upgrading.

File: config/genesis_testnet.json
Comment (gsmachado) [reply config/genesis_testnet.json]: I'd say we need just a bit of GAS for this. Like, 50 GAS to be on the safe side? wdyt?

Comment (steven1227) [reply config/genesis_testnet.json]: Maybe a tiny amount is enough ?

Comment (gsmachado) [reply config/genesis_testnet.json]: Let's set 5 GAS, then. Or, I'm also fine with 1 GAS, if it's enough.

@mialbu will this be more than enough for the initial steps, right?

Comment (mialbu) [reply config/genesis_testnet.json]: Actually, if it's a tiny amount, we need to fund it anyway rather soon already. Then I'm not sure if there's any point in allocating any funds to the bridge in the genesis.

Comment (AnnaShaleva) [reply config/genesis_testnet.json]: Initial transfer from Treasury (where all Bridge funds are stored) requires votes from CN accounts. CN accounts should be supplied with GAS before they can vote. We don't put genesis allocations for CN accounts, they are supposed to bridge initial amount of funds to themselves. So the question is: can we perform these initial bridging to CN accounts with no GAS on the Bridge contract balance? I suppose we can't, that's why we need at least *some* amount for this genesis allocation.

File: config/genesis_testnet.json
Comment (AnnaShaleva) [reply config/genesis_testnet.json]: As discussed, changed to `100,000,000`.

File: contracts/solidity/interfaces/IStub1.sol
Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub1.sol]: @txhsl, agree to use a single stub *implementation*?

File: config/genesis_testnet.json
Comment (mialbu) [reply config/genesis_testnet.json]: Oh okay, I wasn't aware of that. Yes, in that case the bridge needs some initial allocation, absolutely.

There is a minimum deposit amount on the N3 side. It's been 1 Gas so far, but it can be modified as well. So, the initial allocation should be at least that amount, and it also needs to be enough to cover all the fees for the necessary transactions to fund the bridge. If 1 Gas is sufficient for all required transactions, then 1 Gas is sufficient as initial allocation to the bridge.

File: contracts/solidity/interfaces/IStub1.sol
Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub1.sol]: @chenquanyu fixed, please check.

File: config/genesis_testnet.json
Comment (AnnaShaleva) [inline config/genesis_testnet.json]: @txhsl, registration fee is set to 20K GAS, please check it's the right storage item.

Comment (txhsl) [reply config/genesis_testnet.json]: It is the expected key and number value, but for a correct format without error, we may need `0x043c33c1937564800000` here.

File: contracts/solidity/interfaces/IStub1.sol
Comment (txhsl) [reply contracts/solidity/interfaces/IStub1.sol]: > @txhsl, agree to use a single stub _implementation_?

I'm OK with that if it works.

File: config/genesis_testnet.json
Comment (txhsl) [reply config/genesis_testnet.json]: If this value changes before launch, I will let @steven1227 inform here directly.

File: config/genesis_testnet.json
Comment (txhsl) [reply config/genesis_testnet.json]: I'm not sure, but it is somehow different with my deploy code. @chenquanyu Can you check again as well?

File: config/genesis_testnet.json
Comment (AnnaShaleva) [reply config/genesis_testnet.json]: > but for a correct format without error, we may need 0x043c33c1937564800000 here.

Fixed.

File: contracts/solidity/interfaces/IStub1.sol
Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub1.sol]: @chenquanyu could you please check that at least one stub can be correctly upgraded on privnet?

File: config/genesis_testnet.json
Comment (AnnaShaleva) [reply config/genesis_testnet.json]: I used the updated hardhat config from this PR to compile:
```
import { HardhatUserConfig } from "hardhat/config";
import "@nomicfoundation/hardhat-toolbox";

const config: HardhatUserConfig = {
  solidity: {
    version: "0.8.25",
    settings: {
      evmVersion: "shanghai",
      optimizer: {
        enabled: true,
        runs: 500,
      },
    },
  },
  paths: {
    sources: "./solidity",
    tests: "./test",
  }
};

export default config;
```

Then did `npx hardhat compile` and then took the code from `deployedBytecode` field of `go-ethereum/contracts/artifacts/solidity/Treasury.sol/Treasury.json`. Am I doing something wrong?

File: config/genesis_testnet.json
Comment (AnnaShaleva) [reply config/genesis_testnet.json]: 1GAS for BridgeProxy contract is allocated.

File: config/genesis_testnet.json
Comment (txhsl) [reply config/genesis_testnet.json]: No, not. I reset the env and get the same result as this line.

File: contracts/solidity/interfaces/IStub1.sol
Comment (chenquanyu) [reply contracts/solidity/interfaces/IStub1.sol]: > @chenquanyu could you please check that at least one stub can be correctly upgraded on privnet?

I have checked that all stub contracts can be upgraded correctly on privnet. Maybe we should add more stub contracts to meet future needs.

Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub1.sol]: What is the desired number? Do you have exact contract examples that should be added in future (except DKG)?

Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub1.sol]: OK, how about 10 stub contracts? It's a lot, and I think we can fit our future needs into these contracts for sure.

Comment (txhsl) [reply contracts/solidity/interfaces/IStub1.sol]: Yes, it's a lot. Or just from `0x1212000000000000000000000000000000000000` to `0x121200000000000000000000000000000000000f`.

Comment (AnnaShaleva) [reply contracts/solidity/interfaces/IStub1.sol]: Added.

**Comment by AnnaShaleva** [comment]
@txhsl @chenquanyu waiting for you approval before the merge, Roman is on vacation.

**Review by txhsl — APPROVED** [review]

**Pull request merged**
